### ISSUE TITLE
Add-on store: show updates, if any, when focusing add-ons in the installed add-ons tab and remove the updatable add-ons tab

### ISSUE THREAD

**Issue by Adriani90** [nvaccess/nvda#15878]
### Is your feature request related to a problem? Please describe.
Add-ons can have different statuses such as update available, pending removal, incompatible, enabled / disabled. Currently the different statuses of add-ons are splited in different tabs, and some are included in columns. The current UX design is very inconvenient, especially when trying to check if there are updates for incompatible add-ons. You have to change the tab first, then you have to find the checkbox to show incompatible addons and thick it. The current UX design is overcoded with code that in fact is not always needed. The fact that statuses have to be determined on add-ons store tab is inefficient though sometimes still needed. But we should try to minimize statusses that need to be determined on a tab by tab basis.

### Describe the solution you'd like
Users should be incentivized to find updates for their add-ons, especially for incompatible add-ons, as fast as possible. Installed incompatible add-ons should never be filtered out. I propose following change in the first run:
1. Show the update available status on each add-on in the list of installed add-ons in the installed add-ons tab
2. Use speech priority and make NVDA announce the "update available" status when focusing the add-on with arrow keys or with the mouse.
3. Make type of announcement configurable (I) announce with usual pitch rate (II) announce with different pitch rate or (III) announce by playing a sound.
4. List the add-ons with available updates always at the top of the list, similar to what powedit does for example for strings that need a translation update.

Implementing this proposal would make the updatable add-ons tab obsolete and would make it easier for users to find updates for their add-ons.

### Describe alternatives you've considered
Provide notifications (i.e. toast notifications) when updates for add-ons are available.

### Additional context
This proposal can be implemented by using the speech framework implemented in #7599. This use case would be a great start to test this framework in the NVDA core. Unfortuantely its great potential has not been applied yet in practice in the NVDA core to design user experience and achieve more efficiency and less verbosity.
Other points to consider when implementing:
1. Update availability checks should only occur at the time when opening the add-on store, and not instantly while the add-on store is opened. This will make sure the add-ons are not reordered in the list while the user is browsing the list of installed add-ons.
2. Failing update availability checks should be displayed as a column to the coresponding add-on in the list of installed add-ons. This will avoid the case where the user would get promted every time when opening the add-on store in case update checks have failed. The user can hear this while browsing the list of installed add-ons instead.
3. The status incompatible, enabled or disabled should not have any effect on update availability check. But add-ons with status "pending removal" should not check for updates since they will be removed anyway after NVDA restart.
4. The list of available add-ons in the coresponding tab should not include the statusses enabled, disabled, pending removal or update available. So an update check as well as other statusses relevant for installed add-ons are only displayed in the installed add-ons tab.

**Comment by fernando-jose-silva** [comment]
Sorry, I'm not sure I understood the request very well.
But I wouldn't want to lose the functionality of scrolling with the up and down arrows, and only scrolling through add-ons pending updates.
If your request is implemented, to have this functionality, you would have to be able to filter the add-ons with available updates.

**Comment by fernando-jose-silva** [comment]
Perfect, thank you, the add-ons with updates at the top of the list are great for me.

**Comment by wmhn1872265132** [comment]
It is not recommended to do this. For users with poor network or pure offline users, the experience may be very bad:
1. When should local installed add-ons be listed? If you list the local add-ons first, and then change the order if there are add-ons that need to be updated after obtaining the online add-on data, the order of the list may suddenly change during browsing; if you wait for the online data to be updated before loading the local add-on Data, then users with poor network who only want to disable/enable/remove a certain add-on may have to wait a long time.
2. For offline users without network, they can install/update the add-on through external channels, and then disable/enable/remove the add-on through the store. However, after doing this, every time they open the store, they will be prompted Unable to fetch latest add-on data. The user experience will not be particularly good either.

In short, a purely offline part is still needed to manage local add-ons.

**Comment by Adriani90** [comment]
> If you list the local add-ons first, and then change the order if there are add-ons that need to be updated after obtaining the online add-on data, the order of the list may suddenly change during browsing; if you wait for the online data to be updated before loading the local add-on Data, then users with poor network who only want to disable/enable/remove a certain add-on may have to wait a long time.

This can be avoided when you say, data is only fetched when restarting NVDA or when opening the add-on store, but the update availability data is not updated instantly when browsing through the list or as long as the add-on store is opened.

> 2. For offline users without network, they can install/update the add-on through external channels, and then disable/enable/remove the add-on through the store. However, after doing this, every time they open the store, they will be prompted Unable to fetch latest add-on data. The user experience will not be particularly good either.

This can be avoided. Add-ons which have the status disabled or pending removal do not check for update availability data. The data is fetched as soon as you enable the add-on and restart NVDA to take effect.

Also the promt "checking for add-on updates failed" or something similar is not needed for an user. This could be solved with something les intrusive such as a column indicating that the automatic update check failed.

**Comment by wmhn1872265132** [comment]
However, the current Add-on store's Available add-ons tab does not work as you said, so if you implement this issue first, the above situation will still happen.

**Comment by Adriani90** [comment]
> However, the current Add-on store's Available add-ons tab does not work as you said,

The available add-ons tab should not be significantly touched, apart from the status "available update", "enabled / disabled" or "pending removal" which should appear only in the installed add-ons tab, not in the available add-ons tab. In fact then we would have only two tabs where statuses have to be determined on a tab by tab basis, not four tabs.

**Comment by seanbudd** [comment]
>This can be avoided. Add-ons which have the status disabled or pending removal do not check for update availability data. The data is fetched as soon as you enable the add-on and restart NVDA to take effect.

This would reintroduce #15029 and #15568

**Comment by Adriani90** [comment]
Ah thanks Sean for the good remark. In fact, that argument brought by me is indeed even not really relevant as an argument for this request. To be clear, status enabled, disabled or incompatible should not have any effect on update availability check. However, add-ons with status "pending removal" should not check for updates when opening the add-on store.
Failing update checks could be displayed as a column in the list of installed add-ons which is less intrusive and would already solve the concern brought by @wmhn1872265132 above. I updated the main request description accordingly.

**Comment by CyrilleB79** [comment]
I am not in favor of removing the updatable add-ons tab.

I feel it's very convenient to have a dedicated list containing all and only updatable add-ons. Today, the add-on store has one tab per task:
* Installed add-ons tab: "I want to enable/disable/remove an add-on individually"
* Updatable add-ons tab: "I want to check if add-ons have updates and, if any, update all the add-ons which have an update"
* Available add-ons tab: "I want to install one or more new add-ons"
* Incompatible tab: maybe this one is a bit less clear: "I want to review incompatible add-ons and decide to enable them anyway or to remove them (if not used anymore)" Note that this tab should be consulted after having updated updatable incompatible add-ons however.

More generally, removing a tab will lead to the creation of more filters and/or more columns in the list. It was the case in the first prototypes of the add-on store and the resulting GUI has proven to be too complex.

I also disagree with reordering the add-ons in the installed add-ons tab, with updatable add-ons first (such as poedit untranslated strings). This proposal takes only into account the use case of a user who want to update their add-ons. But as written before, the installed add-ons tab is also useful to check what is installed, to remove unused add-ons and to enable/disable them as needed; for this task, a constant alphabetical order is required.

Example:
I have an issue in NVDA which disappears when add-ons are disabled. Thus I want to identify the add-on responsible for this bug. I will disable or reenable add-ons half by half, restart NVDA and test at each step until I find the culprit by dichotomy. During this process, if the order of the add-ons in the list changes due to a new update being available, I will not be able to remember easily which subset of add-ons should be enabled or disabled.

I am in favor of implementing first the toast notification when updates are available. This has proven to be very efficient given the popularity of Add-on Updater add-on and I think that it is expected from many users. If it is not enough for users to update their add-ons, we will see why and what is missing then.

**Comment by sevapopov2** [comment]
I agree with the last comment. Personally, the interface is good for me as it is now!Best regards,Vsevolodhttps://github.com/sevapopov26 дек. 2023 г., в 12:31, Cyrille Bougot ***@***.***> написал(а):﻿
I am not in favor of removing the updatable add-ons tab.
I feel it's very convenient to have a dedicated list containing all and only updatable add-ons. Today, the add-on store has one tab per task:

Installed add-ons tab: "I want to enable/disable/remove an add-on individually"
Updatable add-ons tab: "I want to check if add-ons have updates and, if any, update all the add-ons which have an update"
Available add-ons tab: "I want to install one or more new add-ons"
Incompatible tab: maybe this one is a bit less clear: "I want to review incompatible add-ons and decide to enable them anyway or to remove them (if not used anymore)" Note that this tab should be consulted after having updated updatable incompatible add-ons however.

More generally, removing a tab will lead to the creation of more filters and/or more columns in the list. It was the case in the first prototypes of the add-on store and the resulting GUI has proven to be too complex.
I also disagree with reordering the add-ons in the installed add-ons tab, with updatable add-ons first (such as poedit untranslated strings). This proposal takes only into account the use case of a user who want to update their add-ons. But as written before, the installed add-ons tab is also useful to check what is installed, to remove unused add-ons and to enable/disable them as needed; for this task, a constant alphabetical order is required.
Example:
I have an issue in NVDA which disappears when add-ons are disabled. Thus I want to identify the add-on responsible for this bug. I will disable or reenable add-ons half by half, restart NVDA and test at each step until I find the culprit by dichotomy. During this process, if the order of the add-ons in the list changes due to a new update being available, I will not be able to remember easily which subset of add-ons should be enabled or disabled.
I am in favor of implementing first the toast notification when updates are available. This has proven to be very efficient given the popularity of Add-on Updater add-on and I think that it is expected from many users. If it is not enough for users to update their add-ons, we will see why and what is missing then.

**Comment by nvdaes** [comment]
Cyrille commented: > I am not in favor of removing the updatable add-ons tab.

I strongly agree with Cyrille's comment, super well explained, so nothing to add reading the whole text.

**Comment by lukaszgo1** [comment]
> 2. Use speech priority and make NVDA announce the "update available" status when focusing the add-on with arrow keys or with the mouse.
> 
> 3. Make type of announcement configurable (I) announce with usual pitch rate (II) announce with different pitch rate or (III) announce by playing a sound.


How this should work for sighted people and Braille users? In general we should always try to use standard GUI elements, and only when they're insufficient resort to hacks such as announcing given control differently.

**Comment by XLTechie** [comment]
I agree with @lukaszgo1's points 2 and 3. I am not so sure about point 1, 
as I am not clear on its value if the other points are achieved.

**Comment by CyrilleB79** [comment]
> 1. If there are available updates start Add-on Store in the updateable tab

I tend to agree with this.
It encourages users to update their add-ons as soon as an update is available, which is the main use case.
There may still be use cases where a user does not want to update despite the available update, e.g. dev testing with external install of test version. But such corner use cases should not drive the UX in the GUI.

On the other hand, except before the installation of the first add-on, the add-on store always opens on the same page. If in the future it opens on installed or updatable tab depending on the available updates, will it cause confusion among less experimented users?

> 2. Allow to update / migrate to Add-on Store actions in the context menu from the installed tab

The main goal of this tab is not to update add-ons.
However, if there is no technical limitation to this, there is no reason to prevent the users from updating in this tab.

> 3. Present update statuses in the installed add-ons tab. This makes sense, since other statuses (incompatible, disabled, pending install) are direct effects of user action, so they're of a lesser importance than the fact that given add-on can be updated.

Do you want to add the updatable status, e.g.: status = "Update Available, Disabled (incompatible)".
Or do you want to replace them? E.g.: status = "Update Available" instead of "Disabled"

IMO, none of these two proposal is desirable:
* "Update Available, Disabled (incompatible)" gives too much information for a status information
* Replacing "Disabled" or "Enabled" by "Update available" defeats the main purpose of this tab, i.e. see if add-ons are enabled or disabled. The use case of enabling/disabling add-ons by dichotomy would be broken by this change.

**Comment by CyrilleB79** [comment]
> > 2. Use speech priority and make NVDA announce the "update available" status when focusing the add-on with arrow keys or with the mouse.
> > 3. Make type of announcement configurable (I) announce with usual pitch rate (II) announce with different pitch rate or (III) announce by playing a sound.
> 
> How this should work for sighted people and Braille users? In general we should always try to use standard GUI elements, and only when they're insufficient resort to hacks such as announcing given control differently.

I agree with @lukaszgo1 comment.
IMO earcons and voice changing should not be used in such specific use cases, but in much broader use cases such as:
* voice context (content vs structure or screen reader message)
* voice change upon specific text formatting (e.g.: bold, red) or structure (e.g. headings, links)
* replace role/state by sounds

**Comment by lukaszgo1** [comment]
> 
> > 2. Allow to update / migrate to Add-on Store actions in the context menu from the installed tab
> 
> The main goal of this tab is not to update add-ons. However, if there is no technical limitation to this, there is no reason to prevent the users from updating in this tab.

Thinking about this some more - these options should also be available from the incompatible add-ons tab.

> 
> > 3. Present update statuses in the installed add-ons tab. This makes sense, since other statuses (incompatible, disabled, pending install) are direct effects of user action, so they're of a lesser importance than the fact that given add-on can be updated.
> 
> Do you want to add the updatable status, e.g.: status = "Update Available, Disabled (incompatible)". Or do you want to replace them? E.g.: status = "Update Available" instead of "Disabled"
> 
> IMO, none of these two proposal is desirable:
> 
> * "Update Available, Disabled (incompatible)" gives too much information for a status information
> 
> * Replacing "Disabled" or "Enabled" by "Update available" defeats the main purpose of this tab, i.e. see if add-ons are enabled or disabled. The use case of enabling/disabling add-ons by dichotomy would be broken by this change.



I do share your concerns to some extend. I thought about presenting statuses like this 'status = "Update Available, Disabled (incompatible)"'. Do you have an alternative proposals which would ensure that users are aware that there is an update which can be installed by executing appropriate option  from the context menu?

**Comment by CyrilleB79** [comment]
> I do share your concerns to some extend. I thought about presenting statuses like this 'status = "Update Available, Disabled (incompatible)"'. Do you have an alternative proposals which would ensure that users are aware that there is an update which can be installed by executing appropriate option from the context menu?

No I have not.

And actually, I think it's not necessary since, as I wrote before, updating add-on is not one of the main task in the Installed add-ons tab. There is already the Updatable add-ons tab dedicated to this usage. To me, the possibility to update from Installed add-ons tab should be just seen as an available bonus feature.

I think that a toast notification is enough to notify users of updates being available.

What is presented in each tab should be limited to information useful to the task(s) for which the tab is dedicated. That's the reason why there are various tabs in place of a unique oververbose tab with all possible information and many filtering controls.

**Comment by Adriani90** [comment]
@CyrilleB79 wrote:
> the installed add-ons tab is also useful to check what is installed, to remove unused add-ons and to enable/disable them as needed; for this task, a constant alphabetical order is required.

This could be acomplished by the user manually as soon as sorting by a certain column would be implemented.

> I am not in favor of removing the updatable add-ons tab.
I feel it's very convenient to have a dedicated list containing all and only updatable add-ons.

I see your point after reading all this discussion. But at least in this updatable add-ons tab the incompatible add-ons should not be filtered out by default.

> Incompatible tab: maybe this one is a bit less clear:

Yes and it is confusing. I would rather go for a coresponding filter in the installed add-ons tab instead of a separate tab but there is another issue for this already. In fact we don't really need a column which says "incompatible" when browsing through the list, the filter "show only incompatible add-ons" would already give the same convenience like the separate tab and you don't have to move back and forth between tabs which currently even resets your filters by the way.

Regarding user experience, for update availability checks, I actually agree with points 2 and 3 brought by @lukaszgo1 https://github.com/nvaccess/nvda/issues/15878#issuecomment-1842654816.

@lukaszgo1 wrote:
> How this should work for sighted people and Braille users?

The addons with "update available" chould be displayed in a different color on the screen, documenting it properly in the user guide. For braille users this could be indicated as an alert notification while browsing the list, for which you can set a timeout.

@CyrilleB79 
> "Update Available, Disabled (incompatible)" gives too much information for a status information

It depends on how you manage statuses: 
* incompatible can be handled by a filter, no column needed
* Enabled / disabled can remain a column
* Update available: either column or a more elegant solution with #7599 in mind.

> IMO earcons and voice changing should not be used in such specific use cases, but in much broader use cases

In general I agree with you, but implementing earcons on broader use cases didn't happen for a very long time. i still think this is a good use case to start with and see how users perceive it. if it is not accepted at all, then at least we know how it works and how it feels.

> There is already the Updatable add-ons tab dedicated to this usage. To me, the possibility to update from Installed add-ons tab should be just seen as an available bonus feature.
I think that a toast notification is enough to notify users of updates being available.

I agree that a toast notification would make things better anyway, at least users would get the information of updates available without having to perform any actions which is incentive enough to update add-ons. That's why I proposed this as an alternative.

**Comment by seanbudd** [comment]
Given significant feedback from the community that discourages this change, I am closing as won't fix
